### Limits. The Number $e$ as a Limit
We know what limits are, what they mean and how they relate to our tasks. Let's explore a special limit. This one arises from economics but we'll see it has applications throughout science because of its unique properties.

Imagine you're saving money in a bank. Every period, you accumulate [interest](https://en.wikipedia.org/wiki/Compound_interest) on your money. Let's say the bank is very generous and gives you **100% interest every year**.

How much money will you have after one year? Let's say you got $\$1$. After one year, you'll get your interest from the bank and you'll have $\$2$. Your money doubled, which was expected.

How about this offer: **50% interest every 6 months**? Will this be the same, better, or worse?
You start with $\$1$. After 6 months, you'll accumulate $50%$ interest to get $\$1,50$. After 6 more months, you'll get $50\%.1,50 = 0,75$, so your total is $2,25$. You got $\$0,25$ more!

Let's try to exploit the scheme and get rich.

**$100/12\%$ every month**
* January: $1 + 1/12$
* February: $(1 + 1/12) * (1 + 1/12)$
* March: $(1 + 1/12) * (1 + 1/12) * (1 + 1/12)$
* ...

We can see a pattern. Every period, we multiply our money by $1 + 1/12$. So, the final sum will be $$\$1.\left(1+\frac{1}{12}\right)^{12} = \$2,61$$

We did even better. This is always true. The more periods, the more money we accumulate. The more money we have, the more interest we accumulate. And that completes the vicious circle known as money saving :).

Let's try other options:

**$100/52\%$ every week**

$\$1.\left(1+\frac{1}{52}\right)^{52} = \$2,69$

**$100/31556926\%$ every second**

$\$1.\left(1+\frac{1}{31556926}\right)^{31556926} = \$2,718$

Well, there's a slight problem to our world domination plans. Even though we accumulate more and more money, we get *diminishing returns*. For 52 periods we got $2,69$, and for more than 3 million periods we only got like $\$0,02$ more. This pattern will continue.

Now we can ask ourselves, what is the maximum profit we can accumulate for a year? To do this, we can ask
$$ \lim_{n \ \rightarrow \infty}\left(1+\frac{1}{n}\right)^n = ? $$

It turns out this is a constant. It is approximately equal to $2,71828182\dots$. Since it's very useful, it's got a name: $e$, or Euler's number (sometimes called Napier's number). The limit above is **the definition of $e$**.

Why is it so useful? Let's go back to the original problem. In compound interest, the extra amount after every period is proportional to the amount at the start of the period. In other words, **the rate of change of a value is proportional to the value**. This pops out everywhere in nature and business. Some examples include radioactive decay (more atoms $\Rightarrow$ more decays), cooling down a cup of tea (the rate of cooling down depends on the temperature difference between the cup and the room), animal population models (more animals $\Rightarrow$ more babies), infection models, and so on.

To quickly verify the value of $e$, calculate the limit as we defined it above.

In [12]:
from collections.abc import Callable
import numpy as np

In [13]:
def calculate_limit_at_infinity(func: Callable):
    """
    Calculates a numerical approximation of the limit of the specified function
    as its parameter goes to infinity.
    Args:
        func (Callable): The function to calculate the limit of.
    Returns:
        zip: A pair of x and f(x) for x from 1 to infinity (random huge number).
    """
    x = 10 ** np.arange(0, 10)
    return zip(x, func(x)) # Return a pair of x and f(x) for x from 1 to infinity 

Now let's calculate the limit for:
$$ \lim_{n \ \rightarrow \infty}\left(1+\frac{1}{n}\right)^n  $$

In [14]:
limits = calculate_limit_at_infinity(lambda x: (1 + 1 / x) ** x)
for limit in limits:
    print("The limit of the function at x =", limit[0], "is", limit[1])

The limit of the function at x = 1 is 2.0
The limit of the function at x = 10 is 2.5937424601000023
The limit of the function at x = 100 is 2.7048138294215285
The limit of the function at x = 1000 is 2.7169239322355936
The limit of the function at x = 10000 is 2.7181459268249255
The limit of the function at x = 100000 is 2.7182682371922975
The limit of the function at x = 1000000 is 2.7182804690957534
The limit of the function at x = 10000000 is 2.7182816941320818
The limit of the function at x = 100000000 is 2.7182817983473577
The limit of the function at x = 1000000000 is 2.7182820520115603


Can you notice the convergence?

So, the limit for $ \lim_{n \ \rightarrow \infty}\left(1+\frac{1}{n}\right)^n  $ is $2.71828...$ This is a special number at it is called [$e$](https://en.wikipedia.org/wiki/E_(mathematical_constant)).

$$ e = \sum_{n=0}^{\infty} \frac{1}{n!} = 1 + \frac{1}{1} + \frac{1}{1 \cdot 2} + \frac{1}{1 \cdot 3} + ... = 2.718281828459045$$

The special number $e$ is also used as **the base of the natural logarithm** ($ln$).